design-proposals/klip-61-bootstrapped-table-subscriptions.md

```diff
@@ -0,0 +1,89 @@
+we can guarantee there is no missing data. It will also be more clean and intuitive for users to have to only use one query
+to achieve this.
+
+## What is in scope
+
+- Queries should be supported on `TABLES` with `EMIT ALL` at the end.
+- Bootstrapped subscriptions should retrieve all current rows in the `TABLE` and subscribe to any further changes.
+- Bootstrapped subscriptions should not terminate until the user quits them.
+- Bootstrapped subscriptions will have feature parity with push queries v2, such as filters and projections.
+
+## What is not in scope
+
+- Bootstrapped subscriptions will not be supported on `STREAMS`. Instead, we will throw an informative error that tells the
+user to use `SET 'auto.offset.reset'='earliest';` with a push query in order to get all past data in the `STREAM` while 
+simultaneously subscribing to any future updates. This is because unlike `TABLES` which have a current state that a pull query retrieves,
+there is no current state for a `STREAM`. 
+
+## Value/Return
+
+Adding this feature will ensure that users do not miss any data by trying to implement their own bootstrapped `TABLE` 
+subscription by combining a pull query and a push query. It also reduces the potential complexity by only needing a single
+query to get both the current state of a `TABLE` and any future changes to that `TABLE`.
+
+## Public APIS
+
+Users will be able to add an `EMIT ALL` clause at the end of their push query v2 on a table in order to indicate that they
+want to perform a bootstrapped `TABLE` subscription. The proposed syntax would then look this:
+
+```SQL
+SET 'ksql.query.push.v2.enabled' = 'true';
+SELECT select_expr [, ...]                 
+  FROM table
+  [ WHERE where_condition ]
+  [ AND and_condition ]
+  EMIT ALL;
+```
+
+## Design
+
+For a bootstrapped `TABLE` subscription, we will peek at the latest offset in the ksqlDB output topic using the kafka admin
+console. We will then save this offset and start the pull query, which can now take as long as needed since we have already
+saved the latest offset. Once the pull query finishes and returns the current state, we can now begin the subscription 
+and supply the saved offset as the starting subscription token to the push query in order to receive any future changes.
+
+One limitation of this design is that there is a race condition between getting the latest offsets in the ksqlDB output 
+topic and executing the pull query, since the pull query can possibly see new data. This can result in "time travel" 
+once the subscription starts since we may return older results that the pull query has already returned until we catch up.
+
+## Test plan
+
+- Unit testing.
+- Integration testing in `ksqldb-rest-app` to ensure that modules are working as expected when combined.
+- Functional testing through the rest query validation tests (RQTT) to ensure that we are returning the correct results 
```

Comment:
We might need to extend this functionality a bit.  Currently, there's a flag which effectively says, produce the input either before or after the query is run.  We want a combination if you're really going to test it all.

Reply:
added that we will need to update the RQTT framework in order to properly test this